coprocesses/coprocess.go

```diff
@@ -0,0 +1,148 @@
+package coprocesses
+
+import (
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/joyent/containerpilot/utils"
+)
+
+// Some magic numbers used internally by the coprocess restartLimits
+const (
+	haltRestarts      = -1
+	unlimitedRestarts = -2
+)
+
+// Coprocess configures a process that will run alongside the main process
+type Coprocess struct {
+	Name         string      `mapstructure:"name"`
+	Command      interface{} `mapstructure:"command"`
+	RestartLimit interface{} `mapstructure:"restartLimit"`
+
+	Args           []string
+	restart        bool
+	restartLimit   int
+	restartsRemain int
+	cmd            *exec.Cmd
+	logWriters     []io.WriteCloser
+}
+
+// NewCoprocesses parses json config into an array of Coprocesses
+func NewCoprocesses(raw []interface{}) ([]*Coprocess, error) {
+	var coprocesses []*Coprocess
+	if raw == nil {
+		return coprocesses, nil
+	}
+	var configs []*Coprocess
+	if err := utils.DecodeRaw(raw, &configs); err != nil {
+		return nil, fmt.Errorf("Coprocess configuration error: %v", err)
+	}
+	for _, t := range configs {
+		if err := parseCoprocess(t); err != nil {
+			return nil, err
+		}
+		coprocesses = append(coprocesses, t)
+	}
+	return coprocesses, nil
+}
+
+func parseCoprocess(coprocess *Coprocess) error {
+	args, err := utils.ToStringArray(coprocess.Command)
+	if err != nil {
+		return err
+	}
+	coprocess.Args = args
+	if coprocess.Args == nil || len(coprocess.Args) == 0 {
+		return fmt.Errorf("Coprocess did not provide a command")
+	}
+	cmd := utils.ArgsToCmd(coprocess.Args)
+	coprocess.cmd = cmd
+
+	if coprocess.Name == "" {
+		coprocess.Name = strings.Join(coprocess.Args, " ")
+	}
+
+	switch t := coprocess.RestartLimit.(type) {
+	case string:
+		if t == "unlimited" {
+			coprocess.restart = true
+			coprocess.restartLimit = unlimitedRestarts
+			coprocess.restartsRemain = unlimitedRestarts
+		} else {
+			return fmt.Errorf("`restartLimit` field only takes positive integers or `unlimited`")
+		}
+	case int:
```

Comment:
Also, for consistency - we should support strings with integer contents like we do for stopTimeout, etc...